distributed/tests/test_worker_state_machine.py

```diff
@@ -245,60 +249,235 @@ def test_executefailure_to_dict():
+            ("gather-dependencies", b.address, {"x"}),
+            ("x", "fetch", "flight", "flight", {}),
+            ("request-dep", b.address, {"x"}),
+            ("busy-gather", b.address, {"x"}),
+            ("x", "flight", "fetch", "fetch", {}),
+            ("x", "update-who-has", [b.address], []),  # Called Scheduler.who_has
+            ("x", "fetch", "missing", "missing", {}),
+        ],
+        # There may be a round of find_missing() after this
+        strict=False,
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_self_denounce_missing_data(c, s, a):
+    x = c.submit(inc, 1, key="x")
+    await x
+
+    # Manually wipe x from the worker
+    # (Using an endpoint like remove-replicas would inform the scheduler).
+    del a.data["x"]
+    del a.tasks["x"]
+    a.validate_state()
+
+    y = c.submit(inc, x, key="y")
+    # The scheduler tries computing y, but a responds that x is not available.
+    # The scheduler kicks off the computation of x and then y from scratch.
+    assert await y == 3
+
+    assert_story(
+        a.log,
+        [
+            ("x", "compute-task", "released"),
+            ("x", "released", "waiting", "waiting", {"x": "ready"}),
+            ("x", "waiting", "ready", "ready", {"x": "executing"}),
+            ("x", "ready", "executing", "executing", {}),
+            ("x", "put-in-memory"),
+            ("x", "executing", "memory", "memory", {}),
+            ("y", "compute-task", "released"),
+            ("x", "ensure-task-exists", "released"),
+            ("y", "released", "waiting", "waiting", {"x": "fetch"}),
+            ("x", "released", "fetch", "released", {"x": "missing"}),
+            ("x", "released", "missing", "missing", {}),
+            ("free-keys", ("y",)),
+            ("y", "release-key"),
+            (
+                "y",
+                "waiting",
+                "released",
+                "released",
+                {"x": "released", "y": "forgotten"},
+            ),
+            ("y", "released", "forgotten", "forgotten", {}),
+            ("x", "release-key"),
+            ("x", "missing", "released", "released", {"x": "forgotten"}),
+            ("x", "released", "forgotten", "forgotten", {}),
+            ("x", "compute-task", "released"),
+            ("x", "released", "waiting", "waiting", {"x": "ready"}),
+            ("x", "waiting", "ready", "ready", {"x": "executing"}),
+            ("x", "ready", "executing", "executing", {}),
+            ("x", "put-in-memory"),
+            ("x", "executing", "memory", "memory", {}),
+            ("y", "compute-task", "released"),
+            ("x", "ensure-task-exists", "memory"),
+            ("y", "released", "waiting", "waiting", {"y": "ready"}),
+            ("y", "waiting", "ready", "ready", {"y": "executing"}),
+            ("y", "ready", "executing", "executing", {}),
+            ("y", "put-in-memory"),
+            ("y", "executing", "memory", "memory", {}),
```

Comment:
This log is very long. What is the essential information in here? What are we actually testing?

Reply:
I'll reduce it to the interesting bits.